app/src/lib/stores/helpers/infer-comparison-branch.ts

```diff
@@ -0,0 +1,143 @@
+import { Branch, IAheadBehind } from '../../../models/branch'
+import { PullRequest } from '../../../models/pull-request'
+import { Repository } from '../../../models/repository'
+import { GitHubRepository } from '../../../models/github-repository'
+import { revRange } from '../../git'
+
+/**
+ * Infers which branch to use as the comparison branch
+ *
+ * The branch returned is determined by the following conditions:
+ * 1. Given a pull request -> target branch of PR
+ * 2. Given a forked repository -> default branch on `upstream`
+ * 3. Given a hosted repository -> default branch on `origin`
+ * 4. Fallback -> `master` branch
+ *
+ * @param branches The list of all branches for the repository
+ * @param repository The repository the branch belongs to
+ * @param currentPullRequest The pull request to use for finding the branch
+ * @param currentBranch The branch we want the parent of
+ * @param getAheadBehind function used to calculate ahead/behind
+ */
+export async function inferComparisonBranch(
+  branches: ReadonlyArray<Branch>,
+  repository: Repository,
+  currentPullRequest: PullRequest | null,
+  currentBranch: Branch | null,
```

Comment:
This is for my education only -- can you help me understand what scenario we could be in where currentBranch is null but we still want to infer a comparison branch?

Reply:
There actually isn't one. This exists only to make using this function easier for the caller.That is the reason why I created a discriminated union a bit earlier in the history of this branch. 

The thing is, depending on what we're working with (PR, Fork, Gh Repo, everything else), we need to do different things. Since each helper function accepts different arguments, this function has to accept the union of all of those arguments.

Because of that, I made the 2 parameters that all functions need, `branches` and `repository`, `non-nullable`. I chose to keep `currentPullRequest` and `currentBranch` `nullable` instead of `optional` because the types as they are defined in `app-state` are `nullable`

Does that make sense? Also, I am open to going back to using types, but then I'd be back in the position of needing to come up with good names and the caller would be responsible for knowing the different types this function expects.